### PR TITLE
Allow to control report type - local, cloud or both

### PR DESCRIPTION
**Pull request by Rantzur1992**
Added support for control of report generation, which can be local, cloud or by default, both for supported agent versions.

### CODE REVIEW

**Comment by mstrelex** [comment]
@Rantzur1992 

Why do you have this strange signature instead of your name? `LAPTOP-IB4DPRIG\rantz`

File: TestProject.OpenSDK/Internal/Rest/AgentClient.cs [hunk 1]
```diff
@@ -598,21 +609,42 @@ private ReportSettings InferReportSettings(ReportSettings originalSettings)
...
+            Logger.Trace($"Report Type was set to {inferredReportSettings.ReportType}.");
 
             return inferredReportSettings;
         }
 
+        /// <summary>
+        /// Verify whether the current Agent version supports local report generation.
+        /// </summary>
+        /// <param name="reportType">The request Report Type.</param>
+        /// <throws>AgentConnectionException if the current agent version < <see cref="minLocalReportSupportedVersion"/>.</returns>
```
Comment (mstrelex) [inline TestProject.OpenSDK/Internal/Rest/AgentClient.cs:633]: A bit confusing formulation, perhaps replacing the `<` sign with `older than` would be better.

File: README.md [hunk 2]
```diff
@@ -209,6 +209,23 @@ driver.Report().DisableCommandReports(DriverCommandsFilter.All);
...
+By default, the execution report is uploaded to the cloud, and a local report is created, as an HTML file in a temporary folder.
+
+At the end of execution, the report is uploaded to the cloud and SDK outputs to the console/terminal the path for a local report file:
+
+`Execution Report: {temporary_folder}/report.html`
+
+This behavior can be controlled, by requesting only a `LOCAL` or only a `CLOUD` report.
+
+> When the Agent is offline, and only a _cloud_ report is requested, execution will fail with appropriate message.
+Via a driver constructor:
```
Comment (mstrelex) [inline README.md:223]: Please add a blank line.

File: README.md [hunk 3]
```diff
@@ -209,6 +209,23 @@ driver.Report().DisableCommandReports(DriverCommandsFilter.All);
...
+
+```csharp
+var driver = new ChromeDriver(chromeOptions: new ChromeOptions(), reportType: ReportType.LOCAL);
+```
+
```
Comment (mstrelex) [inline README.md:228]: What about driver builder?

File: TestProject.OpenSDK/Internal/Rest/Messages/SessionResponses/SessionResponse.cs [hunk 4]
```diff
@@ -47,5 +47,10 @@ public abstract class SessionResponse
         /// The current version of the Agent.
         /// </summary>
         public string Version { get; set; }
+
+        /// <summary>
+        /// The local report generated.
+        /// </summary>
+        public string LocalReport { get; }
```
[the remark this answers is not in the document]
Comment (mstrelex) [reply TestProject.OpenSDK/Internal/Rest/Messages/SessionResponses/SessionResponse.cs:54]: And where the path is printed? See https://github.com/testproject-io/java-opensdk/pull/91/commits/85478861e37dd93d0723e6aa720a35d224b33b77#diff-b8310f141bdd5d03ed9a8ea4a6185363318a109e169ec56043b7e92743cffca5R944

**Review by mstrelex — CHANGES_REQUESTED** [review]

**Comment by mstrelex** [comment]
@Rantzur1992 

Please update HISTORY.md

File: TestProject.OpenSDK/Internal/Rest/Messages/SessionResponses/SessionResponse.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (basdijkstra) [reply TestProject.OpenSDK/Internal/Rest/Messages/SessionResponses/SessionResponse.cs:54]: > How does it work when the property in json is camelCase and the C# property is PascalCase?

That's being taken care of here: https://github.com/testproject-io/csharp-opensdk/blob/81c96f4cc1b50de548f564ca23c24a51f36296ce/TestProject.OpenSDK/Internal/Helpers/CustomJsonSerializer.cs#L42. It works for all properties in the `SessionResponse` (and any other Agent response message that should be deserialized into an object).

[the remark this answers is not in the document]
Comment (Rantzur1992) [reply TestProject.OpenSDK/Internal/Rest/Messages/SessionResponses/SessionResponse.cs]: Thanks @basdijkstra 
I added a new commit which adds the print to the path of the local report, if applicable.

File: TestProject.OpenSDK/Internal/Rest/AgentClient.cs (hunk 1 above)
Comment (Rantzur1992) [reply TestProject.OpenSDK/Internal/Rest/AgentClient.cs]: Sure, done.

File: README.md (hunk 2 above)
Comment (Rantzur1992) [reply README.md:224]: done

File: README.md (hunk 3 above)
Comment (Rantzur1992) [reply README.md:229]: DriverBuilder is not implemented in the C# SDK.
I will need to implement it in a separate PR.

Comment (Rantzur1992) [reply README.md:229]: @mstrelex PR #158 addresses the creation of DriverBuilder.
After you merge it I will be able to add the DriverType to DriverBuilder.

**Review by mstrelex — APPROVED** [review]

**Pull request merged**